### PR TITLE
quiche: update QUICHE to commit 5dd7a030209f9a6b5043bebd8ac3ee54f18d1d08

### PR DESCRIPTION
**Pull request by danzh2010**
Signed-off-by: Dan Zhang <danzh@google.com>

Commit Message: Modify QUIC_BUG implementation to log with rate limit like ENVOY_BUG in release mode. Fix a QuicMemSliceSpanImpl life time issue which was exposed by new QUICHE change

Risk Level: low
Testing: added unit tests for QUIC_BUG and QuicMemSliceSpanImpl

### CODE REVIEW

**Comment by moderation** [comment]
You have some CI issues but...

/lgtm deps

**Comment by danzh2010** [comment]
It's ready for review now. PTAL!
Coverage test failure is caused by ProxyProtoFilterChainMatchIntegrationTest.MismatchDirectSourceButMatchSource/IPv4 which is irrelevant to this PR.

File: source/common/quic/platform/quic_mem_slice_span_impl.h [hunk 1]
```diff
@@ -55,22 +60,29 @@ class QuicMemSliceSpanImpl {
 
 private:
   Envoy::Buffer::Instance* buffer_{nullptr};
+  // Nullptr if this span is not constructed with a QuicMemSlice.
+  QuicMemSliceImpl* mem_slice_{nullptr};
```
Comment (wu-bin) [inline source/common/quic/platform/quic_mem_slice_span_impl.h:64]: Can you add some comments to explain the relationship of buffer_ and mem_slice_?  (It seems we only need one of them to be !nullptr at a time, which is easier to understand)

File: source/common/quic/platform/quiche_bug_tracker_impl.h [hunk 2]
```diff
@@ -6,11 +6,46 @@
...
+
+  std::ostringstream& stream() { return stream_; }
+
+private:
+  std::ostringstream stream_;
+  const std::string condition_str_;
+  const std::string bug_name_;
+};
+
+class ScopedDisableExitOnQuicheBug {
```
Comment (wu-bin) [inline source/common/quic/platform/quiche_bug_tracker_impl.h:30]: nit: Clarify (in comments) it is test only, and only one instance can exists at a time.

File: source/common/quic/platform/quiche_bug_tracker_impl.cc [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  if (Envoy::Assert::shouldLogAndInvokeEnvoyBugForEnvoyBugMacroUseOnly(bug_name_)) {
+    ENVOY_LOG_TO_LOGGER(Envoy::Logger::Registry::getLog(Envoy::Logger::Id::envoy_bug), error,
+                        "QUICHE_BUG failure: {}.{}{}", condition_str_,
+                        stream_.eof() ? "" : " Details: ", stream_.str());
+#if !defined(NDEBUG) && !defined(ENVOY_CONFIG_COVERAGE)
+    if (!g_quiche_bug_exit_disabled) {
+      abort();
+    }
+#else
+    Envoy::Assert::invokeEnvoyBugFailureRecordActionForEnvoyBugMacroUseOnly(bug_name_.data());
```
Comment (wu-bin) [inline source/common/quic/platform/quiche_bug_tracker_impl.cc:37]: nit: remove .data()

File: test/common/quic/platform/quic_platform_test.cc [hunk 4]
```diff
@@ -112,9 +118,14 @@ TEST_F(QuicPlatformTest, QuicExpectBug) {
   auto bug = [](const char* error_message) { QUIC_BUG(bug_id) << error_message; };
 
   auto peer_bug = [](const char* error_message) { QUIC_PEER_BUG(bug_id) << error_message; };
-
   EXPECT_QUIC_BUG(bug("bug one is expected"), "bug one");
   EXPECT_QUIC_BUG(bug("bug two is expected"), "bug two");
+#ifdef NDEBUG
+  // The 3rd triggering in release mode should be omitted.
```
Comment (wu-bin) [inline test/common/quic/platform/quic_platform_test.cc:124]: "should not be logged".  If there's a bug counter it should still be incremented.

File: source/common/quic/platform/quic_mem_slice_span_impl.h (hunk 1 above)
Comment (danzh2010) [reply source/common/quic/platform/quic_mem_slice_span_impl.h:64]: done. buffer_ always points to the actually memory, but mem_slice_ is only used in the constructor which takes QuicMemSlice.

File: source/common/quic/platform/quiche_bug_tracker_impl.h (hunk 2 above)
Comment (danzh2010) [reply source/common/quic/platform/quiche_bug_tracker_impl.h:30]: done. But why only one instance can exists at a time?

File: source/common/quic/platform/quiche_bug_tracker_impl.cc (hunk 3 above)
Comment (danzh2010) [reply source/common/quic/platform/quiche_bug_tracker_impl.cc:37]: invokeEnvoyBugFailureRecordActionForEnvoyBugMacroUseOnly() take a char*

File: source/common/quic/platform/quiche_bug_tracker_impl.h (hunk 2 above)
Comment (wu-bin) [reply source/common/quic/platform/quiche_bug_tracker_impl.h:30]: I don't see comment added?

Suppose two threads, each has a ScopedDisableExitOnQuicheBug instance, called S1 and S2, and S1 and S2's lifetime overlaps like:

T=0:   S1 construct
T=1:                                S2 construct
T=2:   S1 destruct
T=3:                               S2 destruct

The final value for ScopedDisableExitOnQuicheBug will be true.

Comment (danzh2010) [reply source/common/quic/platform/quiche_bug_tracker_impl.h:30]: Oh, you mean multiple thread. Add comment for real now.

Comment (wu-bin) [reply source/common/quic/platform/quiche_bug_tracker_impl.h:30]: Yes multi thread is the main problem. But that sequence is problematic in single thread too.

Still not seeing comments. LGTM assuming you'll add one:)

**Review by wu-bin — APPROVED** [review]
Thanks Dan. LGTM but please add a comment for ScopedDisableExitOnQuicheBug.

**Comment by alyssawilk** [comment]
(tagging as waiting on a new hash per internal discussion)

**Comment by danzh2010** [comment]
> (tagging as waiting on a new hash per internal discussion)

Updated to later commit. PTAL

File: bazel/repository_locations.bzl [hunk 5]
```diff
@@ -808,12 +808,12 @@ REPOSITORY_LOCATIONS_SPEC = dict(
...
         project_url = "https://quiche.googlesource.com/quiche",
-        version = "aee86fb6ffce85e884a6f613ae5f47ce2c6b1e23",
-        sha256 = "a1b2c0cdd53a3a932db991bee736c6df20912c3d8070be9fbb4152575837cbf9",
+        version = "5dd7a030209f9a6b5043bebd8ac3ee54f18d1d08",
+        sha256 = "306342cb35cb9d8baea079c7b924b0133c53cbf182b251655e589d3b4604dc41",
         # Static snapshot of https://quiche.googlesource.com/quiche/+archive/{version}.tar.gz
         urls = ["https://storage.googleapis.com/quiche-envoy-integration/{version}.tar.gz"],
         use_category = ["dataplane_core"],
-        release_date = "2021-06-02",
+        release_date = "2021-07-13",
```
Comment (moderation) [inline bazel/repository_locations.bzl:816]: I _think_ the date for this commit is `2021-07-16` as per https://quiche.googlesource.com/quiche/+/5dd7a030209f9a6b5043bebd8ac3ee54f18d1d08 but it's a little difficult to navigate `googlesource.com`. As this isn't a Github repo I don't think we check this programatically

**Review by alyssawilk — APPROVED** [review]
LGTM modulo outstanding comments

File: bazel/repository_locations.bzl (hunk 5 above)
Comment (danzh2010) [reply bazel/repository_locations.bzl]: fixed

**Comment by alyssawilk** [comment]
@moderation I think this needs one more LGTM?

**Pull request merged**